core/market/src/protocol/discovery.rs

```diff
@@ -40,26 +42,76 @@ pub(super) struct OfferHandlers {
 pub struct DiscoveryImpl {
     identity: Arc<dyn IdentityApi>,
 
+    offer_queue: Mutex<Vec<SubscriptionId>>,
+    unsub_queue: Mutex<Vec<SubscriptionId>>,
+
     offer_handlers: Mutex<OfferHandlers>,
     get_local_offers_handler: HandlerSlot<RetrieveOffers>,
     offer_unsubscribe_handler: HandlerSlot<UnsubscribedOffersBcast>,
 }
 
 impl Discovery {
-    /// Broadcasts Offers to other nodes in network. Connected nodes will
-    /// get call to function bound at `OfferBcast`.
     pub async fn bcast_offers(&self, offer_ids: Vec<SubscriptionId>) -> Result<(), DiscoveryError> {
         if offer_ids.is_empty() {
             return Ok(());
         }
-        let default_id = self.default_identity().await?;
+        // When there are 0 items in the queue we should schedule a send job.
+        let must_schedule = {
+            let mut queue = self.inner.offer_queue.lock().await;
+            let result = queue.len() == 0;
+
+            queue.append(&mut offer_ids.clone());
+            result
+        };
+        log::trace!(
+            "bcast_offers done appending {} offers. must_schedule={}",
+            offer_ids.len(),
+            must_schedule
+        );
+
+        if must_schedule {
+            let myself = self.clone();
+            let _ = Arbiter::spawn(async move {
+                // Sleep to collect multiple offers to send
+                delay_for(Duration::from_secs(5)).await;
+                myself.send_bcast_offers().await;
+            });
+        }
+        Ok(())
+    }
+
+    /// Broadcasts Offers to other nodes in network. Connected nodes will
+    /// get call to function bound at `OfferBcast`.
+    async fn send_bcast_offers(&self) -> () {
+        // `...offer_queue` MUST be empty to trigger the sending again
+        let offer_ids: Vec<SubscriptionId> =
+            self.inner.offer_queue.lock().await.drain(0..).collect();
+
+        // Should never happen, but just to be certain.
+        if offer_ids.is_empty() {
+            return ();
+        }
+
+        let default_id = match self.default_identity().await {
+            Ok(id) => id,
+            Err(e) => {
+                log::error!(
+                    "Error getting default identity, not sending bcast. error={:?}",
+                    e
+                );
+                return;
+            }
+        };
+        log::debug!("Broadcasting offers. count={}", offer_ids.len());
         let bcast_msg = SendBroadcastMessage::new(OffersBcast { offer_ids });
 
         // TODO: We shouldn't use send_as. Put identity inside broadcasted message instead.
-        let _ = bus::service(local_net::BUS_ID)
+        if let Err(e) = bus::service(local_net::BUS_ID)
             .send_as(default_id, bcast_msg) // TODO: should we send as our (default) identity?
-            .await?;
-        Ok(())
+            .await
+        {
+            log::error!("Error sending bcast, skipping... error={:?}", e);
```

Comment:
shouldn't we put offer ids back into queue here?

Reply:
i like the idea! but it has the same "problem" as above